server/commands/write.go

```diff
@@ -60,9 +60,37 @@ func (c *WriteCommand) validateTuplesets(ctx context.Context, req *openfgapb.Wri
 	}
 
 	for _, tk := range writes {
-		if _, err := tupleUtils.ValidateTuple(ctx, c.datastore, store, modelID, tk, dbCallsCounter); err != nil {
+		tupleUserset, err := tupleUtils.ValidateTuple(ctx, c.datastore, store, modelID, tk, dbCallsCounter)
+		if err != nil {
 			return serverErrors.HandleTupleValidateError(err)
 		}
+		switch usType := tupleUserset.Userset.(type) {
+		case *openfgapb.Userset_This:
+			continue // no need to check on Direct Relationship
+		case *openfgapb.Userset_Intersection:
+			isDirect := isDirectIntersection(usType, tk)
+			if !isDirect {
+				return serverErrors.HandleTupleValidateError(&tupleUtils.IndirectWriteError{Reason: "Attempting to write directly to an indirect only relationship", TupleKey: tk})
```

Comment:
Seems `Reason` can be a variable.

Reply:
Ah good point -- I lifted these directly out of the functions that were originally returning bool, but that'll clean it up nicely (I don't foresee bespoke messages, but if we want that in the future all we need to do is add additional variables)